src/msca/metrics/main.py

```diff
@@ -0,0 +1,475 @@
+        self, 
+        obs_values: np.ndarray, 
+        pred_values: np.ndarray, 
+        weight_values: np.ndarray
+    ) -> float:
+        """
+        Calculate metric for single set of values.
+        
+        Parameters
+        ----------
+        obs_values : np.ndarray
+            Observed values
+        pred_values : np.ndarray
+            Predicted values
+        weight_values : np.ndarray
+            Sample weights
+            
+        Returns
+        -------
+        float
+            Calculated metric value
+        """
+        match self:
+            case ErrorMetric.ROOT_MEAN_SQUARED_ERROR:
+                mse_value = metrics.mean_squared_error(
+                    y_true=obs_values,
+                    y_pred=pred_values,
+                    sample_weight=weight_values
+                )
+                return np.sqrt(mse_value)
+            case ErrorMetric.MEAN_ABSOLUTE_ERROR:
+                return metrics.mean_absolute_error(
+                    y_true=obs_values,
+                    y_pred=pred_values,
+                    sample_weight=weight_values
+                )
+            case ErrorMetric.MEAN_SQUARED_ERROR:
+                return metrics.mean_squared_error(
+                    y_true=obs_values,
+                    y_pred=pred_values,
+                    sample_weight=weight_values
+                )
+            case ErrorMetric.MEAN_ABSOLUTE_PERCENTAGE_ERROR:
+                return metrics.mean_absolute_percentage_error(
+                    y_true=obs_values,
+                    y_pred=pred_values,
+                    sample_weight=weight_values
+                )
+            case ErrorMetric.MEDIAN_ABSOLUTE_ERROR:
+                return metrics.median_absolute_error(
+                    y_true=obs_values,
+                    y_pred=pred_values,
+                    sample_weight=weight_values
+                )
+            case _:
+                raise ValueError(f"Unsupported metric type: {self}")
+    
+    def _eval_grouped(
+        self, 
+        data: pd.DataFrame, 
+        obs: str, 
+        pred: str, 
+        weights: str, 
+        groupby: List[str]
+    ) -> pd.DataFrame:
+        """
+        Calculate error metrics for each group in the DataFrame.
+        
+        Parameters
+        ----------
+        data : pd.DataFrame
+            Input DataFrame
+        obs : str
+            Observed values column name
+        pred : str
+            Predicted values column name
+        weights : str
+            Weights column name
+        groupby : List[str]
+            Grouping column names
+            
+        Returns
+        -------
+        pd.DataFrame
+            DataFrame with groupby columns and calculated metric column
+        """
+        df = data.copy()
+        metric_column_name = f"{self.value}_metric"
+        
+        def calculate_group_metric(group):
+            """Calculate metric for a single group."""
+            try:
```

Comment:
Let's not use `try, except` here, and directly use `_eval_single` in the `apply` below. It will simplify a lot and I think it will be ok if we let it directly fail rather than return `nan`. If we really want to return `nan`, this logic should be build in `_eval_single` rather than here.